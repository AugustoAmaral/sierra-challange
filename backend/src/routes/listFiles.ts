import { t, type Context } from "elysia";
import { readdirSync, statSync } from "fs";
import { join } from "path";
import type { ListFilesResponse, ErrorResponse } from "../types";
import {
  formatFileSize,
  extractUUID,
  getOriginalName,
  extractText,
  getExtension,
} from "../utils/fileUtils";
import {
  UPLOAD_DIR,
  SEARCHABLE_EXTENSIONS,
  DEFAULT_LIMIT,
  DEFAULT_OFFSET,
} from "../config/constants";

interface ListFilesQuery {
  textSearch?: string;
  content?: boolean;
  limit?: number;
  offset?: number;
}

export async function handleListFiles({
  query,
  set,
}: {
  query: ListFilesQuery;
  set: Context["set"];
}): Promise<ListFilesResponse | ErrorResponse> {
  try {
    const {
      textSearch = "",
      content = false,
      limit = DEFAULT_LIMIT,
      offset = DEFAULT_OFFSET,
    } = query;
    const hasFilters = (textSearch && textSearch.trim() !== "") || content;
    let totalSize = 0;

    const allFiles = readdirSync(UPLOAD_DIR);
    const totalFiles = allFiles.length;
    const filesWithMetadata = allFiles
      .map((filename) => {
        const filePath = join(UPLOAD_DIR, filename);
        const stats = statSync(filePath);
        const extension = getExtension(filename);
        const id = extractUUID(filename);
        const nameWithoutExt = getOriginalName(filename);

        totalSize += stats.size;

        return {
          id,
          name: nameWithoutExt,
          displayName: nameWithoutExt + `.${extension}`,
          size: stats.size,
          sizeFormatted: formatFileSize(stats.size),
          createdAt: stats.birthtime.toISOString(),
          extension,
          filePath,
        };
      })
      .sort(
        (a, b) =>
          new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
      );

    let filteredFiles = filesWithMetadata;
    if (content) {
      filteredFiles = filteredFiles.filter((file) =>
        SEARCHABLE_EXTENSIONS.includes(file.extension.toLowerCase())
      );

      if (textSearch && textSearch.trim() !== "") {
        const searchTerm = textSearch.toLowerCase();

        const filesWithContent = await Promise.all(
          filteredFiles.map(async (file) => {
            const content = await extractText(file.filePath);
            return {
              file,
              content: content.toLowerCase(),
            };
          })
        );

        filteredFiles = filesWithContent
          .filter(({ content }) => content.includes(searchTerm))
          .map(({ file }) => file);
      }
    } else if (hasFilters) {
      const searchTerm = textSearch.toLowerCase();
      filteredFiles = filteredFiles.filter((file) =>
        file.name.toLowerCase().includes(searchTerm)
      );
    }

    let filteredCount = null,
      filteredSize = null;
    if (hasFilters) {
      filteredCount = filteredFiles.length;
      filteredSize = filteredFiles.reduce((sum, file) => sum + file.size, 0);
    }

    const paginatedFiles = filteredFiles.slice(offset, offset + limit);

    return {
      success: true,
      files: paginatedFiles.map(({ filePath, ...item }) => ({
        ...item,
        url: `/api/files/${item.name}${item.id}.${item.extension}`,
      })),
      metadata: {
        total: totalFiles,
        totalSize,
        totalSizeFormatted: formatFileSize(totalSize),
        filteredFiles: filteredCount,
        filteredSize,
        filteredSizeFormatted:
          filteredSize !== null ? formatFileSize(filteredSize) : null,
        limit,
        offset,
      },
    };
  } catch (error) {
    console.error("List files error:", error);
    set.status = 500;
    return {
      success: false,
      error: "Failed to list files",
      code: "LIST_ERROR",
    } as ErrorResponse;
  }
}

export const handleListFilesDocumentation = {
  query: t.Object({
    textSearch: t.Optional(
      t.String({
        description: "Search term for filename or content",
      })
    ),
    content: t.Optional(
      t.Boolean({
        description:
          "If true, search only in file contents and filter by searchable types",
      })
    ),
    limit: t.Optional(
      t.Number({
        description: "Number of files to return per page",
        default: DEFAULT_LIMIT,
      })
    ),
    offset: t.Optional(
      t.Number({
        description: "Number of files to skip",
        default: DEFAULT_OFFSET,
      })
    ),
  }),
  detail: {
    summary: "List uploaded files",
    tags: ["files", "list"],
  },
  response: {
    200: t.Object({
      success: t.Boolean(),
      files: t.Array(
        t.Object({
          id: t.String(),
          name: t.String(),
          displayName: t.String(),
          size: t.Number(),
          sizeFormatted: t.String(),
          createdAt: t.String(),
          extension: t.String(),
          url: t.String(),
        })
      ),
      metadata: t.Object({
        total: t.Number(),
        totalSize: t.Number(),
        totalSizeFormatted: t.String(),
        filteredFiles: t.Nullable(t.Number()),
        filteredSize: t.Nullable(t.Number()),
        filteredSizeFormatted: t.Nullable(t.String()),
        limit: t.Number(),
        offset: t.Number(),
      }),
    }),
    500: t.Object({
      success: t.Literal(false),
      error: t.String(),
      code: t.String(),
    }),
  },
};
