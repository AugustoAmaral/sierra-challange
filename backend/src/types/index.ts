export interface UploadedFile {
  id: string;
  originalName: string;
  filename: string;
  size: number;
  sizeFormatted: string;
  mimeType: string;
  uploadedAt: string;
  url: string;
}

export interface UploadResponse {
  success: boolean;
  message: string;
  files: UploadedFile[];
}

export interface ErrorResponse {
  success: false;
  error: string;
  code: string;
}

export interface FileListItem {
  id: string;
  name: string;
  displayName: string;
  size: number;
  sizeFormatted: string;
  createdAt: string;
  extension: string;
  url: string;
}

export interface ListFilesResponse {
  success: boolean;
  files: FileListItem[];
  metadata: {
    total: number;
    totalSize: number;
    totalSizeFormatted: string;
    filteredFiles: number | null;
    filteredSize: number | null;
    filteredSizeFormatted: string | null;
    limit: number;
    offset: number;
  };
}
