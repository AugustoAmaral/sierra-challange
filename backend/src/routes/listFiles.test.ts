import { describe, expect, it, beforeEach } from "bun:test";
import { writeFile } from "fs/promises";
import { join } from "path";
import { handleListFiles, handleListFilesDocumentation } from "./listFiles";
import { ListFilesResponse } from "../types";

const TEST_UPLOAD_DIR = join(process.cwd(), "test_uploads");

async function createTestFile(filename: string, content: string) {
  await writeFile(join(TEST_UPLOAD_DIR, filename), content);
}
async function createSampleFiles() {
  const testFiles = [
    {
      name: "report0542f4c4-192b-4eaa-9857-91e276088878.pdf",
      content: "Test PDF content",
    },
    {
      name: "documentabc12345-1234-4abc-8abc-123456789abc.txt",
      content: "Test document content with searchable text",
    },
    {
      name: "notes12345678-1234-4123-8123-123456789012.md",
      content: "# Test Markdown\nSome notes here",
    },
  ];

  for (const file of testFiles) {
    await createTestFile(file.name, file.content);
    await new Promise((r) => setTimeout(r, 10)); // Ensure different timestamps
  }
}

describe("listFiles route", () => {
  it("should have proper documentation structure", () => {
    expect(handleListFilesDocumentation.query).toBeDefined();
    expect(handleListFilesDocumentation.detail).toBeDefined();
    expect(handleListFilesDocumentation.response).toBeDefined();
    expect(handleListFilesDocumentation.response[200]).toBeDefined();
    expect(handleListFilesDocumentation.response[500]).toBeDefined();
  });

  it("should include pagination parameters in documentation", () => {
    const querySchema = handleListFilesDocumentation.query;
    expect(querySchema).toBeDefined();
  });

  describe("handleListFiles", () => {
    it("it should return an empty file list when no files exist", async () => {
      const result = await handleListFiles({ query: {}, set: { status: 200 } });

      expect(result).toEqual({
        success: true,
        files: [],
        metadata: {
          total: 0,
          totalSize: 0,
          totalSizeFormatted: "0.00 B",
          filteredFiles: null,
          filteredSize: null,
          filteredSizeFormatted: null,
          limit: 20,
          offset: 0,
        },
      });
    });

    it("should list existing files with correct metadata", async () => {
      await createSampleFiles();

      const result = (await handleListFiles({
        query: {},
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(3);

      const file = result.files[0];
      expect(file).toMatchObject({
        displayName: "notes.md",
        extension: "md",
        id: "12345678-1234-4123-8123-123456789012",
        name: "notes",
        size: 31,
        sizeFormatted: "31.00 B",
        url: "/api/file/12345678-1234-4123-8123-123456789012",
      });
      expect(file.createdAt).toMatch(
        /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z/
      );

      expect(result.metadata.total).toMatchSnapshot();
    });

    it("should sort files by creation date (most recent first)", async () => {
      await createSampleFiles();
      const result = (await handleListFiles({
        query: {},
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(3);

      const file1Date = new Date(result.files[0].createdAt);
      const file2Date = new Date(result.files[1].createdAt);
      expect(file1Date.getTime()).toBeGreaterThanOrEqual(file2Date.getTime());
    });
  });

  describe("Pagination", () => {
    beforeEach(async () => {
      for (let i = 1; i <= 5; i++) {
        await createTestFile(
          `file${i}aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa.txt`,
          `content${i}`
        );
        await new Promise((r) => setTimeout(r, 10)); //Ensure different timestamps
      }
    });

    it("should limit the number of files returned", async () => {
      const result = (await handleListFiles({
        query: { limit: 3 },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(3);
      expect(result.metadata).toMatchSnapshot();
    });

    it("should apply offset correctly", async () => {
      const result = (await handleListFiles({
        query: { offset: 2, limit: 2 },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(2);
      expect(result.files[0].name).toBe("file3");
      expect(result.files[1].name).toBe("file2");
      expect(result.metadata).toMatchSnapshot();
    });

    it("should use default values when limit and offset are not provided", async () => {
      const result = (await handleListFiles({
        query: {},
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.metadata).toMatchSnapshot();
    });
  });

  describe("Search by file name", () => {
    beforeEach(async () => {
      await createTestFile(
        "document550e8400-e29b-41d4-a716-446655440000.pdf",
        "content"
      );
      await createTestFile(
        "image550e8400-e29b-41d4-a716-446655440000.jpg",
        "content"
      );
      await createTestFile(
        "report550e8400-e29b-41d4-a716-446655440000.txt",
        "content"
      );
    });

    it("should filter files by name (case insensitive)", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "DOC" },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(1);
      expect(result.files[0].name).toBe("document");
    });

    it("should return an empty list when no files match the search", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "nonexistent" },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(0);
      expect(result.metadata.filteredFiles).toBe(0);
      expect(result.metadata.filteredSize).toBe(0);
    });

    it("should not ignore whitespace in the search", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "document" },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(1);
      expect(result.files[0].name).toBe("document");
    });
  });

  describe("Content Search", () => {
    beforeEach(async () => {
      for (let i = 0; i <= 4; i++) {
        await createTestFile(
          `file${i}aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa.txt`,
          ["alpha", "beta", "gamma", "delta", "epsilon"][i]
        );
        await new Promise((r) => setTimeout(r, 10)); // Ensure different timestamps
      }
      await createTestFile(
        `imageaaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa.jpg`,
        "binarycontent"
      );
    });

    it("should filter only files with searchable extensions when content=true", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "a", content: true },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(4);
      expect(result.files.map((f) => f.extension)).not.toContain("jpg");
      expect(result.files.map((f) => f.name)).toMatchSnapshot();
    });

    it("should search inside the content when textSearch and content are provided", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "ta", content: true },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(2);
      expect(result.files.map(({ name }) => name)).toMatchSnapshot();
    });

    it("should be case insensitive when searching content", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "GAMMA", content: true },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(1);
    });

    it("should return empty list when no file content matches", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "zeta", content: true },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(0);
    });

    it("should return all files when textSearch is empty with content=true", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "", content: true },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(5); // Apenas os 5 arquivos txt
    });

    it("should return the correct metadata for content search results", async () => {
      const result = (await handleListFiles({
        query: { textSearch: "a", content: true },
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.metadata).toMatchSnapshot();
    });
  });

  describe("Edge cases", () => {
    it("should handle files without extensions", async () => {
      await createTestFile(
        "noextension550e8400-e29b-41d4-a716-446655440000",
        "content"
      );

      const result = (await handleListFiles({
        query: {},
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(1);
      expect(result.files[0].extension).toBe("");
      expect(result.files[0].displayName).toBe("noextension.");
    });

    it("should handle very long file names", async () => {
      const longName = "a".repeat(200);
      await createTestFile(
        `${longName}550e8400-e29b-41d4-a716-446655440000.txt`,
        "content"
      );

      const result = (await handleListFiles({
        query: {},
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files).toHaveLength(1);
      expect(result.files[0].name).toBe(longName);
    });

    it("should calculate URL correctly for different extensions", async () => {
      await createTestFile(
        "test550e8400-e29b-41d4-a716-446655440000.PDF",
        "content"
      );
      const result = (await handleListFiles({
        query: {},
        set: { status: 200 },
      })) as ListFilesResponse;

      expect(result.success).toBe(true);
      expect(result.files[0].url).toBe(
        "/api/file/550e8400-e29b-41d4-a716-446655440000"
      );
    });
  });
});
