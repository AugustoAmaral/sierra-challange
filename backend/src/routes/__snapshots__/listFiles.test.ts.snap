// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`listFiles route handleListFiles should list existing files with correct metadata 1`] = `3`;

exports[`listFiles route Pagination should limit the number of files returned 1`] = `
{
  "filteredFiles": null,
  "filteredSize": null,
  "filteredSizeFormatted": null,
  "limit": 3,
  "offset": 0,
  "total": 5,
  "totalSize": 40,
  "totalSizeFormatted": "40.00 B",
}
`;

exports[`listFiles route Pagination should apply offset correctly 1`] = `
{
  "filteredFiles": null,
  "filteredSize": null,
  "filteredSizeFormatted": null,
  "limit": 2,
  "offset": 2,
  "total": 5,
  "totalSize": 40,
  "totalSizeFormatted": "40.00 B",
}
`;

exports[`listFiles route Pagination should use default values when limit and offset are not provided 1`] = `
{
  "filteredFiles": null,
  "filteredSize": null,
  "filteredSizeFormatted": null,
  "limit": 20,
  "offset": 0,
  "total": 5,
  "totalSize": 40,
  "totalSizeFormatted": "40.00 B",
}
`;

exports[`listFiles route Content Search should filter only files with searchable extensions when content=true 1`] = `
[
  "file3",
  "file2",
  "file1",
  "file0",
]
`;

exports[`listFiles route Content Search should search inside the content when textSearch and content are provided 1`] = `
[
  "file3",
  "file1",
]
`;

exports[`listFiles route Content Search should return the correct metadata for content search results 1`] = `
{
  "filteredFiles": 4,
  "filteredSize": 19,
  "filteredSizeFormatted": "19.00 B",
  "limit": 20,
  "offset": 0,
  "total": 6,
  "totalSize": 39,
  "totalSizeFormatted": "39.00 B",
}
`;
